crates/derive/src/types/payload.rs

```diff
@@ -121,30 +121,30 @@ impl L2ExecutionPayloadEnvelope {
     pub fn to_l2_block_ref(&self, rollup_config: &RollupConfig) -> Result<L2BlockInfo> {
         let L2ExecutionPayloadEnvelope { execution_payload, .. } = self;
 
-        let (l1_origin, sequence_number) = if execution_payload.block_number ==
-            rollup_config.genesis.l2.number
-        {
-            if execution_payload.block_hash != rollup_config.genesis.l2.hash {
-                anyhow::bail!("Invalid genesis hash");
-            }
-            (rollup_config.genesis.l1, 0)
-        } else {
-            if execution_payload.transactions.is_empty() {
-                anyhow::bail!(
-                    "L2 block is missing L1 info deposit transaction, block hash: {}",
-                    execution_payload.block_hash
-                );
-            }
-            let tx = OpTxEnvelope::decode(&mut execution_payload.transactions[0].as_ref())
-                .map_err(|e| anyhow::anyhow!(e))?;
+        let (l1_origin, sequence_number) =
+            if execution_payload.block_number == rollup_config.genesis.l2.number {
+                if execution_payload.block_hash != rollup_config.genesis.l2.hash {
+                    anyhow::bail!("Invalid genesis hash");
+                }
+                (rollup_config.genesis.l1, 0)
+            } else {
+                if execution_payload.transactions.is_empty() {
+                    anyhow::bail!(
+                        "L2 block is missing L1 info deposit transaction, block hash: {}",
+                        execution_payload.block_hash
+                    );
+                }
 
-            let OpTxEnvelope::Deposit(tx) = tx else {
-                anyhow::bail!("First payload transaction has unexpected type: {:?}", tx.tx_type());
-            };
+                let tx = TxDeposit::decode(&mut execution_payload.transactions[0][1..].as_ref())
+                    .map_err(|e| anyhow::anyhow!(e))?;
```

Comment:
I really don't like that I had to do this to strip the `0x7e` deposit prefix @clabby, I think there's an easier way to do this, but the issue is we check if the bytes are deposit transactions based on their first byte and then I think we should be decoding the deposit tx here. Maybe this should just be refactored into a `TxDeposit::decode_with_prefix()` method that gracefully strips the prefix? Curious as to your thoughts here.

Reply:
You can also use `OpTxEnvelope::decode_2718` (I think `OpTxEnvelope::decode` does the same), which will decode a `OpTxEnvelope::TxDeposit(TxDeposit)`. Just need some extra matching.

We should do this - need to ensure that the first tx is always a deposit.

Reply:
So unfortunately `TxDeposit::decode` does not support this, it reads the header first in it's decoding and that's why the unit tests strip the `0x7e` type prefix. e.g. https://github.com/clabby/op-alloy/blob/refcell/consensus-port/crates/op-consensus/src/transaction/optimism.rs#L177

Reply:
For now, I've added a manual deposit tx type check. Will write up a ticket to fix this.

Reply:
Added a ticket here https://github.com/ethereum-optimism/kona/issues/147